File: libs/core/microphone.ts
/**
 * Events and data from sensors
 */
//% color=#C90072 weight=99
namespace input {
    /**
     * gets the level of loudness from 0 (silent) to 255 (loud)
     */
    //% blockId="loudness"
    //% block="Loudness"
    //% deprecated=true
    export function loudness(): number {
        let value = 0
        let max = pins.analogReadPin(AnalogPin.MIC)
        let min = max
        for (let index = 0; index < 32; index++) {
            value = pins.analogReadPin(AnalogPin.MIC)
            if (value > max) {
                max = value
            } else if (value < min) {
                min = value
            }
        }
        value = Math.floor((max - min) / 4)
        return value
    }
}

File: libs/core/microphone.cpp
#include "pxt.h"

namespace input {

#if MICROBIT_CODAL
// Forward declaration of soundLevelCodal from microphone library
int soundLevelCodal();
#endif

/**
* Reads the loudness through the microphone from 0 (silent) to 255 (loud)
*/
//% help=input/sound-level
//% blockId=soundLevel block="sound level"
//% parts="microphone"
//% weight=34 blockGap=8
//% group="Sound"
int soundLevel() {
#if MICROBIT_CODAL
    return soundLevelCodal();
#else
    int min = 1023;
    int max = 0;

    for (int i = 0; i < 32; i++) {
        int level = uBit.io.MICROPHONE.getAnalogValue();
        if (level > max) {
            max = level;
        }
        if (level < min) {
            min = level;
        }
        uBit.sleep(5); // Add a small delay to allow the analog input to settle
    }

    int range = max - min + 0.5;
    int level = floor(range / 4); // Divide by 4 to get a value between 0 and 255

    return level;
#endif
}

}